HDDS-11511. Introduce metrics in deletion services of OM

Tejaskriya opened the pull request:
## What changes were proposed in this pull request?

The following metrics are added in OM's deletingServices:
KeyDeleting
No of keys processed for delete since last restart
No of keys sent to scm for delete and response(success/fail) since last restart

DirDeleting
No of dirs/empty dirs processed for delete since last restart
No of subfiles processed for delete since last restart

Dirpurge
No of dirs purged since last restart
No of subfiles purged since last restart

Keypurge
No of keys purged since last restart

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-11511

## How was this patch tested?

Tested manually in docker at the jmx endpoint `http://localhost:9874/jmx?qry=Hadoop:service=OzoneManager,name=DeletingServiceMetrics` :
```
{
  "beans" : [ {
    "name" : "Hadoop:service=OzoneManager,name=DeletingServiceMetrics",
    "modelerType" : "DeletingServiceMetrics",
    "tag.Context" : "ozone",
    "tag.Hostname" : "7c4d4db221d4",
    "NumDirsPurged" : 10,
    "NumDirsSentForPurge" : 10,
    "NumKeysProcessed" : 0,
    "NumKeysPurged" : 0,
    "NumKeysSentForPurge" : 0,
    "NumSubDirsMovedToDeletedDirTable" : 9,
    "NumSubDirsSentForPurge" : 0,
    "NumSubFilesMovedToDeletedTable" : 3,
    "NumSubFilesSentForPurge" : 3
  } ]
}
```

Tejaskriya commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/DeletingServiceMetrics.java:
```diff
@@ -0,0 +1,316 @@
+  private MutableGaugeLong numDirPurgedInLatestRequest;
+  @Metric("No. of subFiles purged in latest request")
+  private MutableGaugeLong numSubKeysPurgedInLatestRequest;
+
```
numDirPurged is counting the directories purged, What count do we want to add?

Tejaskriya commented:
@ashishkumar50 thank you for the reviews, I have addressed the comments, could you please review the update?

ashishkumar50 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMDirectoriesPurgeResponseWithFSO.java:
```diff
@@ -64,18 +65,21 @@ public class OMDirectoriesPurgeResponseWithFSO extends OmKeyResponse {
       boolean isRatisEnabled, @Nonnull BucketLayout bucketLayout,
       Map<Pair<String, String>, OmBucketInfo> volBucketInfoMap,
-      SnapshotInfo fromSnapshotInfo, Map<String, OmKeyInfo> openKeyInfoMap) {
+      SnapshotInfo fromSnapshotInfo, Map<String, OmKeyInfo> openKeyInfoMap, DeletingServiceMetrics metrics) {
```
You can do it in request class only instead of passing one more argument here.

Tejaskriya replied:
I have moved it to request class now

ashishkumar50 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMDirectoriesPurgeRequestWithFSO.java:
```diff
@@ -168,7 +175,16 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager, TermIn
+        deletingServiceMetrics.setNumSubKeysPurgedInLatestRequest(numSubFilesDeleted);
+        if (path.hasDeletedDir()) {
+          deletingServiceMetrics.incrNumDirPurged(1);
+          deletingServiceMetrics.setNumDirPurgedInLatestRequest(1);
```
If there are multiple paths in request, this will just set 1 or 0..Either you can remove `NumDirPurgedInLatestRequest` metrics or do it after the for loop and set count correctly.

ashishkumar50 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMDirectoriesPurgeRequestWithFSO.java:
```diff
@@ -168,7 +175,16 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager, TermIn
           }
         }
+        deletingServiceMetrics.incrNumSubKeysPurged(numSubFilesDeleted);
+        deletingServiceMetrics.setNumSubKeysPurgedInLatestRequest(numSubFilesDeleted);
```
This set should be done after loop?

Tejaskriya commented:
@ashishkumar50 I have addressed all the comments, thank you for the review.
@errose28 could you please review the patch?

sumitagrawl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/OpenKeyCleanupService.java:
```diff
@@ -228,11 +228,15 @@ public BackgroundTaskResult call() throws Exception {
-            numOpenKeys, numHsyncKeys, Time.monotonicNow() - startTime);
+            numOpenKeys, numHsyncKeys, duration);
       }
+      ozoneManager.getDeletionMetrics().setOpenKeyCleanupIterationMetrics(runCount.get(), startTime,
```
already metrics are present for cleanup, this is not required

sumitagrawl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java:
```diff
@@ -220,8 +225,15 @@ public BackgroundTaskResult call() {
-                pendingKeysDeletion.getKeysToModify(), null, expectedPreviousSnapshotId);
+                pendingKeysDeletion.getKeysToModify(), null, expectedPreviousSnapshotId, true);
             deletedKeyCount.addAndGet(delCount);
+            metrics.setIterationKeyRunCount(run);
```
we do not need capture run count, start time.

sumitagrawl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java:
```diff
@@ -220,8 +225,15 @@ public BackgroundTaskResult call() {
+            metrics.setIterationKeyRunCount(run);
+            metrics.setIterationKeyStartTime(startTime);
+            metrics.setIterationKeyDuration(Time.monotonicNow() - startTime);
+            metrics.setIterationKeysProcessed(keyBlocksList.size());
```
do not need capture iteration information, as its present as part of info log.

sumitagrawl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java:
```diff
@@ -220,8 +225,15 @@ public BackgroundTaskResult call() {
+            metrics.setIterationKeyDuration(Time.monotonicNow() - startTime);
+            metrics.setIterationKeysProcessed(keyBlocksList.size());
+            metrics.setIterationKeysSentForPurge(delCount);
+            metrics.incrNumKeysProcessed(keyBlocksList.size());
```
Check for naming, blocks and keys separately.

Tejaskriya replied:
KeyBlockList has the list of `BlockGroup` objects, where each `BlockGroup` object has (key, blocklist for the key). Hence the `keyBlocksList.size()` would give the number of keys that were handled, that is why I have named it as key

Tejaskriya replied:
`NumKeysProcessed` is count of all keys checked (having both type of results from scm), but `NumKeysSentForPurge` is having only the successfully deleted count, they are different

sumitagrawl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java:
```diff
@@ -91,12 +93,13 @@ public AbstractKeyDeletingService(String serviceName, long interval,
       KeyManager manager,
       HashMap<String, RepeatedOmKeyInfo> keysToModify,
-      String snapTableKey, UUID expectedPreviousSnapshotId) throws IOException {
+      String snapTableKey, UUID expectedPreviousSnapshotId, boolean shouldUpdateMetrics) throws IOException {
```
shouldUpdateMetrics --> isSnapshotDelete

sumitagrawl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java:
```diff
@@ -128,6 +131,12 @@ protected int processKeyDeletes(List<BlockGroup> keyBlocksList,
+      if (shouldUpdateMetrics) {
+        metrics.setIterationKeysDeletionRequest(blockDeletionResults.size());
+        metrics.setIterationKeysDeleteSuccess(delCount);
+        metrics.incrNumKeysDeletionRequest(blockDeletionResults.size());
```
do not need these metrics, caller have metrics of blockCount and keys

sumitagrawl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java:
```diff
@@ -220,8 +225,15 @@ public BackgroundTaskResult call() {
+            metrics.setIterationKeyDuration(Time.monotonicNow() - startTime);
+            metrics.setIterationKeysProcessed(keyBlocksList.size());
+            metrics.setIterationKeysSentForPurge(delCount);
+            metrics.incrNumKeysProcessed(keyBlocksList.size());
```
numKeysProcessed --> number of keys operated

Tejaskriya replied:
NumKeysOperated might sound confusing. As we often use "processed" in the metrics, I think keeping it as processed would be better. Would that be fine? Any other suggestion also would work

sumitagrawl replied:
Same as above, Considering this, NumKeysProcessed and NumKeysSentForPurge is same, so we do not need keep 2 metrics, can have only one, better deletedKeyCount can be moved to Metrics representing same data.

sumitagrawl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/DeletingServiceMetrics.java:
```diff
@@ -0,0 +1,336 @@
+  @Metric("Total no. of sub-files sent for purge")
+  private MutableGaugeLong numSubFilesSentForPurge;
+
+  public void incrNumDirDeleted(long dirDel) {
```
method name and field param should match, else will give wrong info

sumitagrawl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/DeletingServiceMetrics.java:
```diff
@@ -0,0 +1,336 @@
+    setIterationDirDuration(duration);
+    setIterationDirDeleted(dirDel);
+    setIterationSubDirDeleted(subdirDel);
+    setIterationFilesSentForPurge(filesMove);
```
iteration level merics not required as its logged and overall metrics is there and information can be derived from that.

sumitagrawl reviewed (commented):
Plz check for review comments

errose28 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/DeletingServiceMetrics.java:
```diff
@@ -0,0 +1,336 @@
+  public void setDirectoryDeletionIterationMetrics(long runcount, long startTime, long duration,
+                                                   long dirDel, long subdirDel,
+                                                   long filesMove, long subdirMove) {
+    setIterationDirRunCount(runcount);
```
I don't see anything wrong with capturing run count. It has been useful in the past to detect when the service has completed a run after a restart, and can be divided over past time intervals to get average time taken per run.

Tejaskriya commented:
@sumitagrawl @ashishkumar50 @errose28  I have updated the PR to have only restart metrics and removed the iteration specific ones.
If we determine that iteration specific metrics are needed uniformly across all the services, I can raise a different PR for adding it in all roles OM, SCM and DN.
The aim of this PR is focused to add the missing metrics. Could you please review the updated PR?

sumitagrawl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java:
```diff
@@ -456,6 +460,7 @@ public long optimizeDirDeletesAndSubmitRequest(long remainNum,
               " totalRunCount: {}",
           dirNum, subdirDelNum, subFileNum, (subDirNum - subdirDelNum),
           Time.monotonicNow() - startTime, rnCnt);
+      metrics.incrementDirectoryDeletionTotalMetrics(dirNum + subdirDelNum, subdirMoved, subFileNum);
```
validate if {deletedDirsCount, movedDirsCount, movedFilesCount} as variable, PurgeDir metrics {NumSubDirectoriesMoved,  NumSubFilesMoved, NumDirPurged} and Metrics here {dirDel, dirMove, fileMove} are same value, else this will create confusion between metrics, and also with logs.

Tejaskriya replied:
Thanks for the review! I have verified the numbers using integration tests and updated the PR with some checks in the tests.

Tejaskriya commented:
@sumitagrawl @ashishkumar50 When a delete is performed with skipTrash, the logs and metrics show the exact number of deletes as the user perceives. But in the case we don't skipTrash, due to a checkpoint directory created which is also deleted, the count of deleted directories is +1 -- which is reflected in both the pre-existing logs and metrics added in this PR.

All comments have been addressed. Can you please review the updated PR?

sumitagrawl reviewed (approved):
LGTM
